### PR TITLE
add intial `variant_alternative` implementation (#161 http://en.cppre…

### PR DESCRIPTION
**Pull request by artemp**
…ference.com/w/cpp/utility/variant/variant_alternative)

### CODE REVIEW

File: test/t/variant_alternative.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    using type_1 = mapbox::util::variant_alternative<1, variant_type>::type;
+    using type_2 = mapbox::util::variant_alternative<2, variant_type>::type;
+
+    std::cerr << typeid(type_0).name() << std::endl;
+    std::cerr << typeid(type_1).name() << std::endl;
+    std::cerr << typeid(type_2).name() << std::endl;
+    constexpr bool check_0 = std::is_same<int, type_0>::value;
+    constexpr bool check_1 = std::is_same<double, type_1>::value;
+    constexpr bool check_2 = std::is_same<std::string, type_2>::value;
+    CHECK(check_0);
```
Comment (daniel-j-h) [inline test/t/variant_alternative.cpp]: I think you forgot `check_{1,2}` here

File: include/mapbox/variant.hpp [hunk 2]
```diff
@@ -1023,6 +1023,32 @@ ResultType const& get_unchecked(T const& var)
...
+    using type = typename std::conditional<(sizeof...(Types) == Index),
+                                           First,
+                                           typename variant_alternative<Index, Types...>::type>::type;
+};
+
+template <std::size_t Index>
+struct variant_alternative<Index>
+{
+    using type = void;
+};
```
Comment (daniel-j-h) [inline include/mapbox/variant.hpp]: Why do we need this with `type = void`?

**Comment by daniel-j-h** [comment]
Also if we want to be C++17 compatible'ish we need other functionality (like `variant_size`), too, no?

http://en.cppreference.com/w/cpp/utility/variant

**Comment by artemp** [comment]
@daniel-j-h - thanks, I'll take a look and update tomorrow.

File: test/t/variant_alternative.cpp (hunk 1 above)
Comment (artemp) [reply test/t/variant_alternative.cpp]: :+1:

Comment (artemp) [reply test/t/variant_alternative.cpp]: fixed in 4b98c485cf7d74691f7921145054641daa66936e

File: include/mapbox/variant.hpp (hunk 2 above)
Comment (artemp) [reply include/mapbox/variant.hpp]: right, here is alternative implementation 3449d00cf525d8ef98cee0f4a276e2928398a8f9

**Comment by artemp** [comment]
@daniel-j-h 

re: variant_size - yep, good call I'll work on a separate PR. General strategy is to add new features when user(s) ask for it, but `variant_size` compliment `variant_alternative` somewhat so we should add it I think.

**Comment by artemp** [comment]
Note to myself : consider using built-in `__type_pack_element` when available (e.g clang)

**Comment by artemp** [comment]
835ebc19321c6a9696a2072b7fbd5ca3de818860 adds `variant_size` support as a part of this PR.

NOTE: `variant_size_v` is commented out as we're sticking to c++11 for the time being.

**Pull request merged**